### PR TITLE
Fixed Custom SVG Path in Draftail Editor Not Displaying Properly

### LINKED ISSUE #11051: Draftail toolbar items with custom svg path icon not being rendered as svg in block menu
### Issue Summary

If a block style has been registered with an svg path for the control icon, it displays correctly in the toolbar (either pinned or unpinned) but in the block pop-over menu, the path is passed as text into the icon `<div>` element.

For example, a left-align button registered with

```python
control["icon"] = [
        "M 584.19821,168.76226 H 73.024776 C 32.701407,168.76226 0,137.69252 0,99.38113 0,\
        61.069736 32.701407,30 73.024776,30 H 584.19821 c 40.39183,0 73.02477,31.069736 73.02477,\
        69.38113 0,38.31139 -32.63294,69.38113 -73.02477,69.38113 z m 0,\
        555.04902 H 73.024776 C 32.701407,723.81128 0,692.80659 0,654.43015 0,616.05372 32.701407,\
        585.04902 73.024776,585.04902 H 584.19821 c 40.39183,0 73.02477,31.0047 73.02477,69.38113 0,\
        38.37644 -32.63294,69.38113 -73.02477,69.38113 z M 0,376.90564 C 0,338.5292 32.701407,\
        307.52451 73.024776,307.52451 H 949.32209 c 40.39183,0 73.02481,31.00469 73.02481,69.38113 0,\
        38.37644 -32.63298,69.38113 -73.02481,69.38113 H 73.024776 C 32.701407,446.28677 0,415.28208 0,\
        376.90564 Z M 949.32209,1001.3358 H 73.024776 C 32.701407,1001.3358 0,970.3311 0,931.95467 0,\
        893.57823 32.701407,862.57354 73.024776,862.57354 H 949.32209 c 40.39183,0 73.02481,\
        31.00469 73.02481,69.38113 0,38.37643 -32.63298,69.38113 -73.02481,69.38113 z"
    ]
```

On the toolbar:

```html
<button name="left-align"
        class="Draftail-ToolbarButton  Draftail-ToolbarButton--active"
        type="button"
        aria-label="Left align text"
        tabindex="-1"
        data-draftail-balloon="up">
    <svg width="16"
         height="16"
         viewBox="0 0 1024 1024"
         class="Draftail-Icon "
         aria-hidden="true">
        <path d="M 584.19821,168.76226 H 73.024776 C 32.701407,168.76226 0,137.69252 0,99.38113 0,        61.069736 32.701407,30 73.024776,30 H 584.19821 c 40.39183,0 73.02477,31.069736 73.02477,        69.38113 0,38.31139 -32.63294,69.38113 -73.02477,69.38113 z m 0,        555.04902 H 73.024776 C 32.701407,723.81128 0,692.80659 0,654.43015 0,616.05372 32.701407,        585.04902 73.024776,585.04902 H 584.19821 c 40.39183,0 73.02477,31.0047 73.02477,69.38113 0,        38.37644 -32.63294,69.38113 -73.02477,69.38113 z M 0,376.90564 C 0,338.5292 32.701407,        307.52451 73.024776,307.52451 H 949.32209 c 40.39183,0 73.02481,31.00469 73.02481,69.38113 0,        38.37644 -32.63298,69.38113 -73.02481,69.38113 H 73.024776 C 32.701407,446.28677 0,415.28208 0,        376.90564 Z M 949.32209,1001.3358 H 73.024776 C 32.701407,1001.3358 0,970.3311 0,931.95467 0,        893.57823 32.701407,862.57354 73.024776,862.57354 H 949.32209 c 40.39183,0 73.02481,        31.00469 73.02481,69.38113 0,38.37643 -32.63298,69.38113 -73.02481,69.38113 z">
        </path>
    </svg>
</button>
```
image

On the block pop-over menu:

```html
<div role="option"
     aria-selected="false"
     id="downshift-0-item-9"
     class="w-combobox__option w-combobox__option--col2">
    <div class="w-combobox__option-icon">
        M 584.19821,168.76226 H 73.024776 C 32.701407,168.76226 0,137.69252 0,99.38113 0,        61.069736 32.701407,30 73.024776,30 H 584.19821 c 40.39183,0 73.02477,31.069736 73.02477,        69.38113 0,38.31139 -32.63294,69.38113 -73.02477,69.38113 z m 0,        555.04902 H 73.024776 C 32.701407,723.81128 0,692.80659 0,654.43015 0,616.05372 32.701407,        585.04902 73.024776,585.04902 H 584.19821 c 40.39183,0 73.02477,31.0047 73.02477,69.38113 0,        38.37644 -32.63294,69.38113 -73.02477,69.38113 z M 0,376.90564 C 0,338.5292 32.701407,        307.52451 73.024776,307.52451 H 949.32209 c 40.39183,0 73.02481,31.00469 73.02481,69.38113 0,        38.37644 -32.63298,69.38113 -73.02481,69.38113 H 73.024776 C 32.701407,446.28677 0,415.28208 0,        376.90564 Z M 949.32209,1001.3358 H 73.024776 C 32.701407,1001.3358 0,970.3311 0,931.95467 0,        893.57823 32.701407,862.57354 73.024776,862.57354 H 949.32209 c 40.39183,0 73.02481,        31.00469 73.02481,69.38113 0,38.37643 -32.63298,69.38113 -73.02481,69.38113 z
    </div>
    <div class="w-combobox__option-text">Left align text</div>
</div>
```

image

The centre-align, right-align and code-block icons are also registered with svg paths and display the same behaviour.

### Steps to Reproduce

Create a block style in wagtail_hooks. Use the [help-text example](https://docs.wagtail.org/en/stable/extending/extending_draftail.html#creating-new-blocks) from docs, but replace label with icon defined as svg path (question mark):

```python
import wagtail.admin.rich_text.editors.draftail.features as draftail_features
from wagtail.admin.rich_text.converters.html_to_contentstate import BlockElementHandler

@hooks.register('register_rich_text_features')
def register_help_text_feature(features):
    """
    Registering the `help-text` feature, which uses the `help-text` Draft.js block type,
    and is stored as HTML with a `<div class="help-text">` tag.
    """
    feature_name = 'help-text'
    type_ = 'help-text'

    control = {
        'type': type_,
        'icon': ["m 334.22246,308.44491 c 0,-76.23832 61.98413,-138.22245 138.22245,-138.22245 h 69.11123 c 76.23832,0 138.22246,61.98413 138.22246,138.22245 v 7.77502 c 0,47.08202 -23.97296,90.92446 -63.49594,116.19325 l -91.14044,58.52857 c -54.42509,34.98756 -87.25292,95.24391 -87.25292,159.81972 v 3.23958 c 0,38.22715 30.88408,69.11123 69.11123,69.11123 38.22715,0 69.11123,-30.88408 69.11123,-69.11123 v -3.02361 c 0,-17.70975 9.07084,-34.12367 23.75698,-43.62646 L 691.00917,548.8224 C 770.05514,497.85287 818.00105,410.38397 818.00105,316.21993 v -7.77502 C 818.00105,155.75229 694.24876,32 541.55614,32 H 472.44491 C 319.75229,32 196,155.75229 196,308.44491 c 0,38.22715 30.88408,69.11123 69.11123,69.11123 38.22715,0 69.11123,-30.88408 69.11123,-69.11123 z M 507.00053,999.5572 a 86.38904,86.38904 0 1 0 0,-172.77808 86.38904,86.38904 0 1 0 0,172.77808 z"],
        'description': 'Help text',
        # Optionally, we can tell Draftail what element to use when displaying those blocks in the editor.
        'element': 'div',
    }

    features.register_editor_plugin(
        'draftail', feature_name, draftail_features.BlockFeature(control, css={'all': ['help-text.css']})
    )

    features.register_converter_rule('contentstate', feature_name, {
        'from_database_format': {'div[class=help-text]': BlockElementHandler(type_)},
        'to_database_format': {'block_map': {type_: {'element': 'div', 'props': {'class': 'help-text'}}}},
    })
```

Include the feature in your draftail-editor. 
In a rich text editor, check button displays correctly on the toolbar
image

Select the '+' or type '/' for the block popover menu, icon is rendered as path text instead:
image

### Technical details

-   Python version: 3.10.6
-   Django version: 4.2.2
-   Wagtail version: 5.1.1
-   Browser version: Chrome 118

### PR DESCRIPTION
**Pull request by Temidayo32**
This PR resolves #11051. I successfully reproduced the error reported here. The PR resolves the issue by implementing a fix for Wagtail to handle custom SVG paths.  Thanks to @enzedonline for pointing out the issue

msedge_O4k3sxuTYK

_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: docs/extending/extending_draftail.md [hunk 1]
```diff
@@ -68,7 +80,7 @@ For detailed configuration options, head over to the [Draftail documentation](ht
 
 -   The `type` is the only mandatory piece of information.
 -   To display the control in the toolbar, combine `icon`, `label` and `description`.
--   The controls’ `icon` can be a string to use an icon font with CSS classes, say `'icon': 'fas fa-user',`. It can also be an array of strings, to use SVG paths, or SVG symbol references for example `'icon': ['M100 100 H 900 V 900 H 100 Z'],`. The paths need to be set for a 1024x1024 viewbox.
+-   The controls’ `icon` can be a string to use an icon font with CSS classes, say `'icon': 'fas fa-user',`. It can also be a string, to use SVG paths, or SVG symbol references for example `'icon': 'M100 100 H 900 V 900 H 100 Z',`. The paths need to be set for a 1024x1024 viewbox.
```
Comment (lb-) [inline docs/extending/extending_draftail.md:83]: I think we should try to keep the documentation as is here. If you re-read it says that we use the value as a path if it's an array.

File: docs/extending/extending_draftail.md [hunk 2]
```diff
@@ -57,6 +57,18 @@ def register_mark_feature(features):
...
+
+```python
+WAGTAILADMIN_RICH_TEXT_EDITORS = {
+    'default': {
+        'WIDGET': 'wagtail.admin.rich_text.DraftailRichTextArea',
+        'OPTIONS': {
+            'features': ['h2', 'bold', 'italic', 'link', 'document-link', 'mark']
+        }
+    },
+}
```
Comment (lb-) [inline docs/extending/extending_draftail.md:70]: This docs change seems unrelated to this PR / fix, maybe it's best to make this change as a separate PR if you think it's needed.

File: docs/extending/extending_draftail.md [hunk 3]
```diff
@@ -83,6 +95,7 @@ Blocks are nearly as simple as inline styles:
 ```python
 import wagtail.admin.rich_text.editors.draftail.features as draftail_features
 from wagtail.admin.rich_text.converters.html_to_contentstate import BlockElementHandler
+from wagtail import hooks
```
Comment (lb-) [inline docs/extending/extending_draftail.md:98]: Please pull out these documentation changes for additional hooks imports, these can be done in a separate PR.

File: client/src/components/Icon/Icon.tsx [hunk 4]
```diff
@@ -13,13 +13,26 @@ const Icon: React.FunctionComponent<IconProps> = ({
...
-}) => (
-  <>
-    <svg className={`icon icon-${name} ${className || ''}`} aria-hidden="true">
-      <use href={`#icon-${name}`} />
-    </svg>
-    {title && <span className="visuallyhidden">{title}</span>}
-  </>
-);
+}) => {
+  if (name && name.startsWith('M')) {
```
Comment (lb-) [inline client/src/components/Icon/Icon.tsx:17]: If we re-read the docs we should be checking if the `name` is an array, and then adjust the rendering accordingly. 

```suggestion
  if (Array.isArray(name)) {
```

Also, checking if the string starts with a specific character is very risky, any icon may randomly break. E.g. if we ever add a Monkey or Money icon. 🐒🙊

I know most icons are lower case now but just explaining why we would want to avoid this.

File: client/src/components/Icon/Icon.tsx [hunk 5]
```diff
@@ -13,13 +13,26 @@ const Icon: React.FunctionComponent<IconProps> = ({
...
+    return (
+      <svg className={`icon ${className || ''}`} aria-hidden="true">
+        <path d={name} />
```
Comment (lb-) [inline client/src/components/Icon/Icon.tsx:20]:
```suggestion
        <path d={name.join(' ')} />
```

Assuming we have an Array, we need to convert it back to a string. I'm not sure if we should add the space to join but maybe the documentation could be updated with another example.

File: client/src/components/Icon/Icon.tsx [hunk 6]
```diff
@@ -13,13 +13,26 @@ const Icon: React.FunctionComponent<IconProps> = ({
...
+
+  return (
+    <>
+      <svg
+        className={`icon icon-${name} ${className || ''}`}
+        aria-hidden="true"
+      >
+        <use href={`#icon-${name}`} />
+      </svg>
+      {title && <span className="visuallyhidden">{title}</span>}
```
Comment (lb-) [inline client/src/components/Icon/Icon.tsx:33]: We should still ensure we render a title, even for custom icons.

File: client/src/components/Icon/Icon.test.js [hunk 7]
```diff
@@ -11,6 +11,11 @@ describe('Icon', () => {
     expect(shallow(<Icon name="test" />)).toMatchSnapshot();
   });
 
+  it('#name for SVG path data', () => {
+    const svgPathData = 'M10 30 L20 10 L30 30 L40 30 L50 10 L60 30';
```
Comment (lb-) [inline client/src/components/Icon/Icon.test.js:15]:
```suggestion
    const svgPathData = ['M10 30 L20 10 L30 30 L40 30 L50 10 L60 30'];
```

Additionally,  it would be good for this unit test to do more than just a snapshot. Maybe it should check that there is a `path` element with the appropriate attribute.

**Review by lb- — CHANGES_REQUESTED** [review]
I think this is in the right direction - thank you for digging into the React internals. 

However, if the current documentation says 'give us an array and we will render your custom icon' we should align with that in our code. 

Rendering a custom icon based on a single starting character in the sting is fragile and also changes what we currently tell developers should provide us. 

Also, it's possible that upper case M isn't all a path can start with https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Paths#line_commands

File: docs/extending/extending_draftail.md (hunk 1 above)
Comment (Temidayo32) [reply docs/extending/extending_draftail.md:83]: Oh, I see. Alright, I will revert the changes. Thanks

File: docs/extending/extending_draftail.md (hunk 2 above)
Comment (Temidayo32) [reply docs/extending/extending_draftail.md:70]: Right, I should revert this.

File: docs/extending/extending_draftail.md (hunk 3 above)
Comment (Temidayo32) [reply docs/extending/extending_draftail.md:98]: I will work on this.

File: client/src/components/Icon/Icon.tsx (hunk 4 above)
Comment (Temidayo32) [reply client/src/components/Icon/Icon.tsx:17]: I agree with this. Thanks. I will make the necessary change.

File: client/src/components/Icon/Icon.tsx (hunk 5 above)
Comment (Temidayo32) [reply client/src/components/Icon/Icon.tsx:20]: I will test this out and make the correct documentation update.

File: client/src/components/Icon/Icon.tsx (hunk 6 above)
Comment (Temidayo32) [reply client/src/components/Icon/Icon.tsx:33]: Alright. I will look into this.

File: client/src/components/Icon/Icon.test.js (hunk 7 above)
Comment (Temidayo32) [reply client/src/components/Icon/Icon.test.js:15]: I will work on this.

File: client/src/components/Icon/Icon.tsx (hunk 5 above)
Comment (Temidayo32) [reply client/src/components/Icon/Icon.tsx:20]: @lb- I have tried this, it doesn't work. Literally, we are like back to square one with the same problem I am trying to resolve. Like the solution I currently have, it only work if `icon` is directly set to string. 
fHklT4048v

File: client/src/components/Icon/Icon.tsx (hunk 4 above)
Comment (Temidayo32) [reply client/src/components/Icon/Icon.tsx:17]: @lb-  I am not sure why after checking that `name` is `array` and then using `name.join()`. It doesn't work.

File: client/src/components/Icon/Icon.tsx (hunk 5 above)
Comment (lb-) [reply client/src/components/Icon/Icon.tsx:20]: It looks you you're rendering the string as is. Are you sure you are passing in the path as an array via the config?

**Comment by Temidayo32** [comment]
@lb- I have made the changes. Let me know if they are in a satisfactory direction. You might also want to note that to manage the `svgPath` more easily I created it own separate props.

File: client/src/components/Icon/Icon.tsx (hunk 5 above)
Comment (Temidayo32) [reply client/src/components/Icon/Icon.tsx]: I have created a seperate props for `svgPath`, let me know what you think of the corrections I made. This works, and I think it might be easier managing the `svgPath` data by using its own seperate props rather than getting from `name` props.

**Comment by lb-** [comment]
Thanks @Temidayo32  I may not get to this for another week or so - but pinging @thibaudcolas if you have some time to look at this.

**Comment by Temidayo32** [comment]
Thanks @lb-

**Review by lb- — COMMENTED** [review]
@Temidayo32  this had a few issues still and as such I have rebased locally and refined the solution, added unit tests and validated.

I have also incorporated part of the approach from #12418 - thanks @joelwilliam2005

I will force push shortly but have a back up of your original branches if anything goes wrong.

## Approach taken

1. Simplify the changes to React `Icon`, instead of it doing the whole job of accepting a svgPath, it just accepts `children` now and will render those instead of `use` if provided. It also accepts any other SVG attributes (such as `viewBox`).
2. In ComboBox, resolve the logic similar to the other PR by checking `Array.isArray` on the `icon`, then render it as an array of paths. Nice approach @joelwilliam2005 - However, I did this earlier in the file where the other icon logic sits.
3. Finally, lots and lots of extra unit tests, it cannot be understated for everyone that unit tests for things like this are super important. I have refactored the Icon tests to do more testing and less snapshot based + added more clearer tests to the ComboBox elements.

Thanks both for your inputs here, I felt that this seems to be a pain point so it's best to get the fix in.

FYI @thibaudcolas @laymonage  (probably too late for 6.3 but flagging this fix nonetheless)


## Validation screenshots

Screenshot 2024-10-27 at 7 46 34 PM
Screenshot 2024-10-27 at 7 46 03 PM
Screenshot 2024-10-27 at 7 25 50 PM

**Review by lb- — APPROVED** [review]
Approved
